### PR TITLE
Fix error with GroupingPattern in GroupDetectors

### LINKED ISSUE #15749: GroupDetectors "Grouping Pattern" property does not work properly
### Expected behavior
- Group spectra with the workspace indices corresponding to Grouping Pattern i.e
  `GroupingPattern = "1,2"` should return a workspace containing the spectra at the second and third position in the workspace. 
- If we have a workspace that contains `sp-1, sp-3` only and `GroupingPattern = "0,1"` then we should get the same workspace back as an Output.
### Actual behavior
- If we have a workspace that contains `sp-1, sp-3` only and `GroupingPattern = "0,1"` then the algorithm fails to run and a warning is given stating "Spectra number 2 does not exist in workspace"
### Steps to reproduce the behavior
- Load any dataset that has 3 or more spectra using LoadNexus
- In the LoadNexus dialog specify a spectrum list that contains a gap i.e "1,3" or "3,7" etc.
- Bring up the algorithm dialog for `GroupDetectors` with the workspace from LoadNexus as the Input Workspace
- In the "Grouping Pattern" Property enter the same value you did for LoadNexus ("1,3" for example)
- See that the algorithm produces an error complaining about not be able to find the spectra between the two numbers you provided.
### Platforms affected
- All

### PR DESCRIPTION
**Pull request by gemmaguest**
This PR fixes a bug where the GroupingPattern property in the GroupDetectors algorithm does not give the correct results for workspaces where the spectrum number is not 1-based.

**To test:**
Run the following script. Check the output - the last two calls to GroupDetectors should both give the same result (a sum of 19).
```
def printVal(ws, i):
    print "Index=" + str(i) + ", Spectrum=" + str(ws.getSpectrum(i).getSpectrumNo()) + ", Value = " + str(ws.readY(i))

sample = CreateWorkspace(DataX=[1], DataY=[1,2,3,4,5,6,7,8,9,10], NSpec=10)
print "Original workspace:"
for i in range(0,10):
    printVal(sample,i)
print

group1 = GroupDetectors(sample, GroupingPattern="4-5")
print "Grouped original workspace with GroupingPattern='4-5'. This sums rows 4-5 => spectra 5-6 giving a result of 11:"
printVal(group1,0)
print

cropped = CropWorkspace(sample, StartWorkspaceIndex=4)
print "Cropped workspace:"
for i in range(0,6):
    printVal(cropped,i)
print

group2 = GroupDetectors(cropped, GroupingPattern="4-5")
print "Grouped cropped workspace with GroupingPattern='4-5'. This should sum rows 4-5 => spectra 9-10 giving a result of 19 (but in the bug it sums rows 0-1 => spectra 5-6 giving a result of 11):"
printVal(group2,0)
print

group3 = GroupDetectors(cropped, WorkspaceIndexList="4-5")
print "Grouped cropped workspace with WorkspaceIndexList='4-5'. This sums rows 4-5 => spectra 9-10 giving a result of 19:"
printVal(group3,0)
print
```

Fixes #15749 .

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/DataHandling/src/GroupDetectors2.cpp [hunk 1]
```diff
@@ -1276,12 +1284,26 @@ void GroupDetectors2::translateInstructions(const std::string &instructions,
...
+
+/**
+* Translate the PerformIndexOperations processing instructions from a vector
+* into a format usable by GroupDetectors.
+*
+* @param groups : A vector of gorups, each group being a vector of its 0-based
+* spectrum indices
+* @param commands : A stringstream to be filled
+*/
+void convertGroupsToMapFile(std::vector<std::vector<int>> groups,
```
Comment (PranavBahuguna) [inline Framework/DataHandling/src/GroupDetectors2.cpp]: The documentation is missing a `@param` tag for `axis`.

File: Framework/DataHandling/src/GroupDetectors2.cpp [hunk 2]
```diff
@@ -1215,38 +1245,16 @@ void translateRange(const std::string &instructions,
...
-                                            std::stringstream &commands) {
-  // vector of groups, each group being a vector of its spectra
+* Translate the PerformIndexOperations processing instructions into a vector
+*
+* @param instructions : Instructions to translate
+* @return : A vector of groups, each group being a vector of its 0-based
+* spectrum indices
+*/
+std::vector<std::vector<int>>
+translateInstructions(const std::string &instructions) {
```
Comment (PranavBahuguna) [inline Framework/DataHandling/src/GroupDetectors2.cpp]: Why not move this anonymous namespace these functions are defined in to the top of the file so two separate anonymous workspaces aren't required?

**Comment by PranavBahuguna** [comment]
After running the testing instructions, I've found that both the logic and functionality work fine and have no further issues with them. There are some other issues I wish to address however:

- Check the line comments I have added.
- Should we add unit tests checking for this (specifically that summing rows gives the same results for `GroupingPattern` and `WorkspaceIndexList`?

**Comment by PranavBahuguna** [comment]
Everything else looks fine. Just address the last line comment I added and I will mark it with the shipit icon.

File: Framework/DataHandling/src/GroupDetectors2.cpp [hunk 3]
```diff
@@ -1293,6 +1293,7 @@ translateInstructions(const std::string &instructions) {
 *
 * @param groups : A vector of gorups, each group being a vector of its 0-based
 * spectrum indices
+* @param : The spectra axis of the workspace
```
Comment (PranavBahuguna) [inline Framework/DataHandling/src/GroupDetectors2.cpp]: Its '@param axis'

Comment (gemmaguest) [inline Framework/DataHandling/src/GroupDetectors2.cpp]: Thanks, that's done now

**Comment by PranavBahuguna** [comment]
@gemmaguest There is a merge conflict that needs resolving.

**Pull request merged**